osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineBlueprintContainer.cs

```diff
@@ -121,6 +123,47 @@ protected override void Update()
             }
 
             base.Update();
+
+            updateStacking();
+        }
+
+        private void updateStacking()
+        {
+            // because only blueprints of objects which are alive (via pooling) are displayed in the timeline, it's feasible to do this every-update.
+
+            const int stack_offset = 5;
+
+            // after the stack gets this tall, we can presume there is space underneath to draw subsequent blueprints.
+            const int stack_reset_count = 3;
+
+            Stack<double> currentConcurrentObjects = new Stack<double>();
+
+            // Reversing is done to enumerate in order of increasing StartTime.
+            foreach (var b in SelectionBlueprints.Reverse())
+            {
+                while (currentConcurrentObjects.TryPeek(out double stackEndTime))
+                {
+                    if (Precision.AlmostBigger(stackEndTime, b.HitObject.StartTime, 1))
+                        break;
+
+                    currentConcurrentObjects.Pop();
+                }
+
+                b.Y = -(stack_offset * currentConcurrentObjects.Count);
+
+                var bEndTime = b.HitObject.GetEndTime();
+
+                // if the stack gets too high, we should have space below it to display the next batch of objects.
+                // importantly, we only do this if time has incremented, else a stack of hitobjects all at the same time value would start to overlap themselves.
+                if (!currentConcurrentObjects.TryPeek(out double nextStackEndTime) ||
+                    !Precision.AlmostEquals(nextStackEndTime, bEndTime, 1))
```

Comment:
As far as I can tell, @smoogipoo's fail case lies somewhere in here.

* First off, to check if time has incremented since the last stacked HO, the *start* time needs to be checked (as stacking in a single time instant is based on the start time).
* Secondly, only checking against the whole stack's end time is not catching the aforementioned case (because, for example, if the last HO from the stack starts at time 1000 and ends at time 3000, but the one being considered next also starts at time 1000, this check won't catch it). As far as I can tell all previous start times should be checked against here. Probably could do away with the `Stack<double>` and keep a hashset + the stack's end time as a whole?

Reply:
I'm not 100% sure on how to implement what you propose (although it sounds like it may work). I've pushed a fix for the cases mentioned above, either way. @smoogipoo can you give it a test and make sure this fixes whatever breakage you were seeing?

@bdach if you think that your proposal will make readability of this code better, maybe just PR your version and we can compare them.

Reply:
I actually believe that the current version of stacking as you have it is better than what I proposed, and should be correct. I don't have a formal proof or anything, but looking at the invariants of how the stack is manipulated I don't see any flaw in it (and as far as I can tell it is indeed sufficient to only check the last item in the stack when deciding whether to continue stacking or to reset depth to 0, due to the sort by start time).